### ISSUE TITLE
kern build problem 

### ISSUE THREAD

**Issue by jxq96** [project-dune/dune#25]
Althoght my platform meets the requriements of dune(all yes when run dune_req.sh). But I cannot build the dune, actuall failed when build kern. My OS is Ubuntu16.04, kernel version 4.15.0-142. From the compiler error: implicit declaration of function ‘x2apic_wrmsr_fence’, I think the problem is the kernel version. So which kernel should I use 
?

**Comment by jeremyyx** [comment]
I met the same problem. I think we should go back to version 3.0+, Unless we modify the code to be compatible with the new kernel. :(

**Comment by Limaomao821** [comment]
Yes, you have to either go back to older version linux or modify dune source code.  

Some header files are no longer provided in linux.

**Comment by jeremyyx** [comment]
> Yes, you have to either go back to older version linux or modify dune source code.
> 
> Some header files are no longer provided in linux.

I saw that you forked a repo called Loongson Dune, does it work?

**Comment by Limaomao821** [comment]
> > Yes, you have to either go back to older version linux or modify dune source code.
> > Some header files are no longer provided in linux.
> 
> I saw that you forked a repo called Loongson Dune, does it work?

Yes, though may not be perfect, but it does.

In fact, loongson-dune is something my lab classmate is working on. And my project is built upon it.

**Comment by Daniel1993** [comment]
I'm also trying to build Dune in a recent kernels and I found some issues.
For starters the compiler can't find a header named asm/ipi.h.

After looking into the loongson-dune repository I think they do not support x86, at
least when I try to compile it prints:
loongarch/arch.c: In function ‘arch_dump_regs’:
loongarch/arch.c:22:27: error: ‘struct kvm_regs’ has no member named ‘gpr’
   22 |   (unsigned long long)regs.gpr[0],

In fact I need this DAX feature:
https://www.kernel.org/doc/Documentation/filesystems/dax.txt
which I think is only available in recent kernels.
Is there any chance Dune can be supported in more recent kernels?

**Comment by Limaomao821** [comment]
> I'm also trying to build Dune in a recent kernels and I found some issues. For starters the compiler can't find a header named asm/ipi.h.
> 
> After looking into the loongson-dune repository I think they do not support x86, at least when I try to compile it prints: loongarch/arch.c: In function ‘arch_dump_regs’: loongarch/arch.c:22:27: error: ‘struct kvm_regs’ has no member named ‘gpr’ 22 | (unsigned long long)regs.gpr[0],
> 
> In fact I need this DAX feature: https://www.kernel.org/doc/Documentation/filesystems/dax.txt which I think is only available in recent kernels. Is there any chance Dune can be supported in more recent kernels?

Sorry, I did not make it clear about loongson-dune. And yes, loongson-dune is for MIPS and Loongarch architecture ( Loongson is our company's name).

**Comment by YixinSong-e** [comment]
> I'm also trying to build Dune in a recent kernels and I found some issues. For starters the compiler can't find a header named asm/ipi.h.
> 
> After looking into the loongson-dune repository I think they do not support x86, at least when I try to compile it prints: loongarch/arch.c: In function ‘arch_dump_regs’: loongarch/arch.c:22:27: error: ‘struct kvm_regs’ has no member named ‘gpr’ 22 | (unsigned long long)regs.gpr[0],
> 
> In fact I need this DAX feature: https://www.kernel.org/doc/Documentation/filesystems/dax.txt which I think is only available in recent kernels. Is there any chance Dune can be supported in more recent kernels?

you need to change your linux kernel to 4.4.98 and even older.
